### PR TITLE
Feature/rework crypto provider with shared model

### PR DESCRIPTION
**Pull request by Alexey-N-Chernyshov**
### Description of the Change
Implement crytpo provider for shared_model. It is divided into crypto_signer that needs keypair and crypto_verifier that doesn't. Old model crypto_provider is left for CLI and commented with \deprecated.
Also crypto_verifier checks for signatures size.

### Benefits
Crypto_provider is now on shared_model

### Alternate Designs *[optional]*
I have had an issue with inheritance of function with template parameter of shared_model, so I wrote instead of `shared_model::interface::Signable`
```
+      void sign(shared_model::interface::Block &block) const override;
+
+      void sign(shared_model::interface::Query &query) const override;
+
+      void sign(
+          shared_model::interface::Transaction &transaction) const override;
```

### CODE REVIEW

File: irohad/validation/impl/stateless_validator_impl.cpp [hunk 1]
```diff
@@ -20,15 +20,18 @@
 #include <chrono>
 #include <utility>
```
Comment (kamilsa) [inline irohad/validation/impl/stateless_validator_impl.cpp]: Check if we can easily remove old stateless validator. If so, do it and mark [this](https://soramitsu.atlassian.net/browse/IR-1149) task as done

File: shared_model/cryptography/crypto_provider/crypto_verifier.hpp [hunk 2]
```diff
@@ -18,37 +18,50 @@
...
+  }  // namespace interface
+}  // namespace shared_model
 
 namespace shared_model {
   namespace crypto {
     /**
-     * CryptoVerifier - wrapper for generalization verification of cryptographic
+     * CryptoVerifier - interface for generalization verification of cryptographic
      * signatures
-     * @tparam Algorithm - cryptographic algorithm for verification
```
Comment (kamilsa) [inline shared_model/cryptography/crypto_provider/crypto_verifier.hpp]: Return this parameter and remove crypto_verifier_impl

File: test/integration/pipeline/transfer_asset_inter_domain_test.cpp [hunk 3]
```diff
@@ -16,7 +16,9 @@
  */
 
 #include "backend/protobuf/from_old_model.hpp"
+#include "cryptography/crypto_provider/crypto_signer_impl.hpp"
 #include "integration/pipeline/tx_pipeline_integration_test_fixture.hpp"
+#include "backend/protobuf/from_old_model.hpp"
```
Comment (muratovv) [inline test/integration/pipeline/transfer_asset_inter_domain_test.cpp]: 1. This include has written twice. Another on line 18
2. This `include` looks like not allowed here. Ask a question which task is responsible for solving it?

File: test/integration/pipeline/tx_pipeline_integration_test.cpp [hunk 4]
```diff
@@ -19,12 +19,14 @@
...
 #include "builders/protobuf/queries.hpp"
 #include "builders/protobuf/transaction.hpp"
+#include "cryptography/crypto_provider/crypto_signer_impl.hpp"
 #include "cryptography/crypto_provider/crypto_defaults.hpp"
 #include "cryptography/ed25519_sha3_impl/internal/ed25519_impl.hpp"
 #include "datetime/time.hpp"
 #include "framework/integration_framework/integration_test_framework.hpp"
 #include "integration/pipeline/tx_pipeline_integration_test_fixture.hpp"
 #include "model/generators/query_generator.hpp"
+#include "model/sha3_hash.hpp"
```
Comment (muratovv) [inline test/integration/pipeline/tx_pipeline_integration_test.cpp]: sha3 looks doubtful here. We have to use shared model here.

File: test/integration/pipeline/tx_pipeline_integration_test_fixture.cpp [hunk 5]
```diff
@@ -16,9 +16,13 @@
  */
 
 #include "integration/pipeline/tx_pipeline_integration_test_fixture.hpp"
+
 #include <algorithm>
 #include <atomic>
+
 #include "backend/protobuf/from_old_model.hpp"
+#include "cryptography/crypto_provider/crypto_signer.hpp"
+#include "model/sha3_hash.hpp"
```
Comment (muratovv) [inline test/integration/pipeline/tx_pipeline_integration_test_fixture.cpp]: same as before

**Comment by muratovv** [comment]
Before merging the pull request we have to wait until https://github.com/hyperledger/iroha/pull/1041 PR  will be merged. Also, need to use block validator that checks the set of signatures.

File: test/integration/pipeline/tx_pipeline_integration_test.cpp (hunk 4 above)
Comment (Alexey-N-Chernyshov) [reply test/integration/pipeline/tx_pipeline_integration_test.cpp]: There is still old model that uses hash(). Reworking it with shared_model is not related to the PR.

File: test/integration/pipeline/tx_pipeline_integration_test_fixture.cpp (hunk 5 above)
Comment (Alexey-N-Chernyshov) [reply test/integration/pipeline/tx_pipeline_integration_test_fixture.cpp]: sap

File: shared_model/backend/protobuf/transaction.hpp [hunk 6]
```diff
@@ -26,6 +26,7 @@
 #include "backend/protobuf/common_objects/signature.hpp"
 #include "block.pb.h"
 #include "utils/lazy_initializer.hpp"
+#include "utils/polymorphic_wrapper.hpp"
```
Comment (muratovv) [inline shared_model/backend/protobuf/transaction.hpp]: Looks redundant

File: shared_model/builders/protobuf/unsigned_proto.hpp [hunk 7]
```diff
@@ -49,14 +49,11 @@ namespace shared_model {
...
-        auto signedBlob = shared_model::crypto::DefaultCryptoAlgorithmType::sign(
-            shared_model::crypto::Blob(unsigned_.payload()), keypair);
-        iroha::protocol::Signature protosig;
-        protosig.set_pubkey(crypto::toBinaryString(keypair.publicKey()));
-        protosig.set_signature(crypto::toBinaryString(signedBlob));
-        auto *s1 = new Signature(protosig);
-        unsigned_.addSignature(detail::PolymorphicWrapper<Signature>(
-            s1));  // TODO: 05.12.2017 luckychess think about false case
+        auto signedBlob =
+            shared_model::crypto::DefaultCryptoAlgorithmType::sign(
```
Comment (muratovv) [inline shared_model/builders/protobuf/unsigned_proto.hpp]: There is should used a crypto provider with template algorithm type:
```
shared_model::crypto::CryptoSigner<T>::sign
```
where `T` is a parameter of the method.

File: shared_model/cryptography/crypto_provider/crypto_model_signer.hpp [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+  namespace crypto {
+    template <typename Algorithm = CryptoSigner<>>
+    class CryptoModelSigner {
+     public:
+      explicit CryptoModelSigner(const shared_model::crypto::Keypair &keypair);
+
+      virtual ~CryptoModelSigner() = default;
+
+      virtual void sign(shared_model::interface::Block &block) const;
```
Comment (muratovv) [inline shared_model/cryptography/crypto_provider/crypto_model_signer.hpp]: Looks like code duplication here for each model object.
I suggest to rework it with template implementation.

File: shared_model/backend/protobuf/block.hpp [hunk 9]
```diff
@@ -68,15 +68,22 @@ namespace shared_model {
         return *signatures_;
       }
 
-      bool addSignature(
-          const interface::types::SignatureType &signature) override {
-        if (signatures_->count(signature) > 0) {
+      bool addSignature(const crypto::Signed &signed_blob,
+                        const crypto::PublicKey &public_key) override {
+        // if already has such signature
+        if (std::find_if(signatures_->begin(),
```
Comment (muratovv) [inline shared_model/backend/protobuf/block.hpp]: Add todo for reduce duplication later.

File: test/module/shared_model/cryptography/crypto_usage_test.cpp [hunk 10]
```diff
@@ -61,22 +61,31 @@ class CryptoUsageTest : public ::testing::Test {
...
-        shared_model::crypto::toBinaryString(keypair.publicKey()));
-    wrong_signature.set_signature(
-        shared_model::crypto::toBinaryString(signedBlob));
-    signable.addSignature(shared_model::detail::PolymorphicWrapper<
-        shared_model::proto::Signature>(
-        new shared_model::proto::Signature(wrong_signature)));
+    signable.addSignature(signedBlob, keypair.publicKey());
+  }
+
+  template <typename T>
```
Comment (muratovv) [inline test/module/shared_model/cryptography/crypto_usage_test.cpp]: Add documentation for the method.

**Pull request merged**